### DISCUSSION TITLE
How to generate the extrinsic execution proof

### DISCUSSION THREAD

**Discussion by liuchengxu** [paritytech/substrate#10783] [Q&A]
Basically, substrate already supports the block execution proof, which is being used to validate the state transition of parachain block in polkadot.

```rust
// Signal to record the proof before initializing the block
if record_proof.yes() {
    api.record_proof();
}

// Block execution

// Extract the proof after finalizing the block
let proof = self.api.extract_proof();
```

What we want is the more fine-grained extrinsic execution proof in subspace, we use a custom executive (related: https://github.com/paritytech/substrate/issues/10711) for collecting the storage root after initializing block and applying each extrinsic. If any executor, who is responsible for processing the transactions in subspace, kind of the collator in parachain, detects an invalid state transition of an extrinsic based on the collected intermediate storage roots, then the executor will generate a proof for this invalid extrinsic state transaction, which is the extrinsic execution proof.

My initial idea is that perhaps it might work if we refer to the pattern of the block execution proof, which means to generate the extrinsic proof by using `api.record_proof()` before applying the extrinsic and `api.extract_proof()` after applying the extrinsic, for `api.record_proof()` will reset the inner `recorder` to the default that is sort of a reset.

https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/primitives/api/proc-macro/src/impl_runtime_apis.rs#L274-L276

```rust
for xt in extrinsics {
    api.record_proof();
    apply_extrinsic();
    let extrinsic_execution_proof = api.extract_proof();
}
```

This approach partially works until I found that if there are two same kind of extrinsics in a block, the generated proof will be just empty and finally I think it's probably caused by the overlay when retrieving the storage from the state:

https://github.com/paritytech/substrate/blob/a2ae8a5702c782f1de05dae77b39f13c6ddda3a8/primitives/state-machine/src/ext.rs#L185-L189

The reason is that the underlying `ProofRecorder` only exists in the backend not in the overlay, if a storage item already exists in the overlay, then it won't be recorded during the execution of next round, thus the storage proof for the extrinsic execution can't be constructed as expected.

In order to support the extrinsic execution proof, we need to also embed a proof recorder in the overlay, something like that. I'm not sure if that's possible and how many efforts are required in substrate. Collect me if I'm wrong, any inputs are appreciated!

**Comment by bkchr** [comment]
> This approach partially works until I found that if there are two same kind of extrinsics in a block

This should basically fails in the moment you have multiple extrinsics reading and then writing the same storage entry, because the value is then found in the overlay. As you already have found out, accessing the overlay isn't found in the proof, because it doesn't need to. 

For your use case, you would require to commit all the intermediate state into a trie and then generate the proof based on this.

**Reply by liuchengxu** [reply]
>For your use case, you would require to commit all the intermediate state into a trie and then generate the proof based on this.

@bkchr Thanks! Could you please elaborate a bit on this? Any similar example I can follow?

**Reply by bkchr** [reply]
We don't supports this. There is no reason to support this. Each intermediate trie would also have a different trie root. So, I'm not really sure what that should bring you.

TBH, I also don't really understand what you intermediate proofs should "prove".

**Reply by liuchengxu** [reply]
Yeah, totally understandable, it's just a special use case in subspace. Please allow me to explain a bit more on _what these intermediate proofs should "prove"_. Basically, it's the same purpose of the block execution proof already supported in substrate, which is used to validate the state transition of a parachain block by relay chain validators. Similarly, these intermediate proof will also be used for the state transition verification, the different thing is that in polkadot the state transition of each single parachain block will be verified by the relay chain validators, which is kind of the way of ZK rollup to some extent, while subspace works more like the OP rollup which needs to generate a fraud proof only in case of any invalid state transition detected. The core of this fraud proof is basically the extrinsic execution proof so that it can be verified by someone who does not own the whole state. Hope this makes more sense :(

**Reply by bkchr** [reply]
Yeah makes more sense, but when not just use the state proof of the entire block? That should contain all the relevant data? 

Let's say you want to check transaction 5 in the block, so you would apply transaction 1 - 4 using the state proof and then transaction 5 would be checked by your custom logic. By first applying all the other transactions, you make sure the "state" is in the correct state.

**Reply by nazar-pc** [reply]
Executor commits to all intermediate state roots by signing deterministic execution receipt (which includes those intermediate state roots). Hence fraud proof only needs to include one data for one transaction.

Including the whole block means fraud proof transactions will be much larger and heavier, becoming DoS attack vector on farmers.

Our consensus nodes (farmers) are meant to be much more lightweight than executors (somewhat similar to validators in PoS).

**Reply by bkchr** [reply]
Why you collect these proofs while building the block? You could also just apply the offending transaction when you need the proof and collect it while doing this.

**Reply by nazar-pc** [reply]
We need to be able to say on which transaction exactly did executor misbehave. If we only have the final root there is no other way to verify the execution except execute the whole block from start to finish again and we want to avoid that if at all possible.

**Reply by bkchr** [reply]
And what happens when a transaction misbehaves? How can a transaction misbehave?

**Reply by nazar-pc** [reply]
Not transaction, executor. Since transactions are already determined when farmer signs the block, execution is deterministic from there onwards. So if executor end up with unexpected result, they'll be punished. But we need an efficient way to determine what exactly went wrong.

**Reply by bkchr** [reply]
Who will be punished? What are executors? You also still did not said what "went wrong" means.

**Reply by nazar-pc** [reply]
Farmers are consensus nodes, they just produce blocks (without executing transactions), executors is a separate node role on the network that is responsible for transaction execution, they are punished for executing transactions incorrectly (this is what goes wrong).

**Reply by liuchengxu** [reply]
Back to the original problem, thanks to the previous effort on the light client support in substrate, I found that tweaking `execution_proof` might be workable.

https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/network/src/light_client_requests/handler.rs#L165

The obstacle now is the `overlay` is hardcoded to `Default::default()` but we need to pass a proper overlay. My idea is to execute the first n - 1 extrinsics using the block builder to derive the `OverlayeredChanges` for the nth extrinsic and then use this `execution_proof` method. Any way to get the overlay while executing the extrinsics?

https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/client/service/src/client/call_executor.rs#L308-L315

**Reply by liuchengxu** [reply]
With some hacking into substrate, I added the [extrinsic execution proof](https://github.com/subspace/subspace/blob/811a83b72f/cumulus/client/block-builder/src/lib.rs#L276-L326) in our custom block builder , but had trouble to verify this approach does work. The idea is to abuse the way of [`api.record_proof()` and `api.extract_proof()`](https://github.com/subspace/subspace/blob/811a83b72f/cumulus/client/block-builder/src/lib.rs#L231-L274) mentioned in the first comment, which should be able to generate the execution proof for the first extrinsic, and then compare it with one created by `execution_proof` in the test https://github.com/subspace/subspace/blob/811a83b72f/cumulus/client/cirrus-executor/src/processor.rs#L294, only to find these two proof mismatches, a few of storage keys have different values and I can't tell the meaning of these mismatched storage keys(I can only identify the storage key for `:extrinsic_index`) by inspecting `OverlayedChanges`. Can be really helpful if you could shed some light on what can be related and if there is anything logically wrong here. Thank you! @bkchr

**Reply by bkchr** [reply]
@liuchengxu I tried to check out your code, but you are using some patches to some unknown substrate repo etc. Could you make your code compile-able for me?

**Reply by liuchengxu** [reply]
@bkchr Sure! I just pushed a commit https://github.com/subspace/subspace/commit/737ebaf341a0e4757d6313902ab8882d0901c7be to make it compile-able. The forked substrate is https://github.com/subspace/substrate/tree/proof-generation, which is mainly for exposing the `overlay` in `ApiExt` and a lot of debugging prints :P

```bash
$ cd /tmp
$ git clone https://github.com/subspace/subspace --branch intermediate-storage-proof
$ cd subspace
$ (cd cumulus/client/cirrus-executor/ && cargo test proof -- --nocapture)
```

**Reply by liuchengxu** [reply]
>only to find these two proof mismatches, a few of storage keys have different values and I can't tell the meaning of these mismatched storage keys(I can only identify the storage key for :extrinsic_index) by inspecting OverlayedChanges

@bkchr I have figured out this problem. First, the substrate-test-runtime test environment has some deviations to the regular runtime, so different `OverlayedChanges` occurred. After switching to a _more regular_ runtime, `OverlayedChanges` produced using these two approaches are exactly the same, but they still produced a different storage proof and finally it turned out it's the inner records using `HashMap` that is the culprit because `HashMap` does not maintain a deterministic order for the items. Once I replace the `HashMap` using `BTreeMap`, I have two identical storage proofs generated using two methods. 

https://github.com/paritytech/substrate/blob/4cd2cc958f5a078d87e4eae877167c40cf2ebf9e/primitives/state-machine/src/proving_backend.rs#L118

So, my proposal here is:

1. Given that no obvious performance cost IMO, replace `HashMap` with `BTreeMap` for the inner `records` in the favor of determinism.
2. Extend `prove_execution` with an optional `OverlayedChanges` so that we can generate an execution proof for any execution at any stage. https://github.com/paritytech/substrate/blob/f449abc336754c37646a64825576fa06c46760b4/client/service/src/client/call_executor.rs#L293-L298

I'll send a PR(perhaps two PRs) later if no objections. @bkchr

**Reply by bkchr** [reply]
Good that you found it :) 

And sorry for me not having answered.

**Comment by nazar-pc** [comment]
Briefly the idea is that in Subspace we have optimistic execution and fraud proofs. To make fraud proofs efficient for resource-constrained farmers (consensus nodes, they don't execute transactions), fraud proof doesn't cover the whole block, instead it only needs to cover the first transaction which is executed incorrectly in the block.

For this to work we need to collect execution trace (effective state roots after every transaction), then in fraud proof we can provide inputs to a particular transaction and correct state root and farmers can just run that one transaction to verify a fraud proof.

When compared to Polkadot, the biggest differences are:
* instead of executing the whole block (and including inputs for all transactions) only one transaction is executed
* instead of doing this all the time pessimistically farmers only do execution when fraudulent execution is detected by executors (separate node role on the network dedicated to just transaction execution)

**Comment by burdges** [comment]
Can you explain what the dispute covers?  It's some storage even like a node serving the wrong data?  I'd think bad nodes would simply not serve the data when requested.

---

We knew about partial block fraud proofs when we designed substrate and polkadot, well the [fraud proofs](https://arxiv.org/abs/1809.09044v1) paper gave us the erasure coding idea.

We decided slashing failed disputes, and punishing both sides on indeterminate disputes, worked out much simpler than partial block fraud proofs.  We later noticed partial block fraud proofs break too many optimizations, including cryptographic ones like pre-batching Schnorr or Groth16, SnarkPack, etc.

It'll work out differently if you want end users verifying everything of course, like LazyLedger / Celestia or the fraud proofs paper.  We also believe this whole approach fails though, due to a tragedy of the commons, maybe details could alter this conclusion, but overall sounds tricky.

I do advise deeper consideration over whether partial block fraud proofs really do what you think they do.  

All this said..

I do believe greater room for optimism exists in when doing storage work, so yeah maybe another approach makes sense here.  I'd likely focus upon erasure coding but maybe that brings other annoyances.  Also state channels, not sure.

We've only really have one sort of storage and proof in substrate, which fits poorly with many important situations.  As a prime example, any parachain doing zk UTXOs must implement this all themselves.  It's kinda par for the course that if you need to so something interesting then you may require stranger validation code somehow.  We should keep teams who run into this sort of things talking together and with us, maybe some shared abstractions emerge, but who knows..

**Reply by nazar-pc** [reply]
Dispute covers divergence in what is by definition a deterministic execution of the blocks by executors (network role responsible for transaction processing), while blocks themselves are produced by farmers (Proof-of-Archival-Storage consensus nodes).

The reason we can't just do slashing is because farmers (consensus nodes) that need to run computations will be overloaded long before they can do slashing, so fraud proofs need to be as efficient as possible. Farmers are by definition much weaker nodes in terms of processing power, this allows for more decentralized participation.

If you have time (and I don't expect that you do) it might be beneficial to take a look at our whitepaper so that terminology and overall construction makes more sense: https://subspace.network/blog/subspace-network-whitepaper
We do indeed have some unusual things comparing to most Substrate-based chains. We are fine writing custom logic and contributing back to Substrate, forking half of Substrate and modifying a bunch of things downstream is not a very efficient way in terms of maintenance.
